src/libos/src/fs/file_ops/utimes.rs

```diff
@@ -0,0 +1,192 @@
+use crate::util::mem_util::from_user;
+
+use super::time::{timespec_t, timeval_t, OcclumTimeProvider};
+use super::*;
+
+use rcore_fs::dev::TimeProvider;
+
+const UTIME_NOW: i64 = (1i64 << 30) - 1i64;
+const UTIME_OMIT: i64 = (1i64 << 30) - 2i64;
+
+bitflags! {
+    pub struct UtimeFlags: i32 {
+        const AT_SYMLINK_NOFOLLOW = 1 << 8;
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+pub struct utimbuf_t {
+    atime: time_t,
+    mtime: time_t,
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+#[allow(non_camel_case_types)]
+pub struct utimes_timespec_tuple_t {
+    atime: timespec_t,
+    mtime: timespec_t,
+}
+
+impl utimes_timespec_tuple_t {
+    pub fn from_utimbuf_ptr(ptr: *const utimbuf_t) -> Result<utimes_timespec_tuple_t> {
+        from_user::check_ptr(ptr)?;
+        let utimbuf = unsafe { *ptr };
+        let atime = timespec_t::from(utimbuf.atime);
+        atime.validate()?;
+        let mtime = timespec_t::from(utimbuf.mtime);
+        mtime.validate()?;
+        Ok(utimes_timespec_tuple_t { atime, mtime })
+    }
+
+    pub fn from_timeval_ptr(ptr: *const timeval_t) -> Result<utimes_timespec_tuple_t> {
+        from_user::check_array(ptr, 2)?;
+        let atime_ptr = unsafe { ptr.offset(0) };
+        let atime = unsafe { *atime_ptr };
+        let atime = timespec_t::from(atime);
+        atime.validate()?;
+        let mtime_ptr = unsafe { ptr.offset(1) };
+        let mtime = unsafe { *mtime_ptr };
+        let mtime = timespec_t::from(mtime);
+        mtime.validate()?;
+        Ok(utimes_timespec_tuple_t { atime, mtime })
+    }
+
+    pub fn from_timespec_ptr(ptr: *const timespec_t) -> Result<utimes_timespec_tuple_t> {
+        from_user::check_array(ptr, 2)?;
+        let atime_ptr = unsafe { ptr.offset(0) };
+        let atime = unsafe { *atime_ptr };
```

Comment:
Should we do more checks before using those unsafe points?

Reply:
I think the check is sufficient currently...